gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/auth/GCSDelegationTokens.java

```diff
@@ -0,0 +1,384 @@
+                e);
+    }
+  }
+
+  public Text getService() {
+    return service;
+  }
+
+  public URI getUri() {
+    return uri;
+  }
+
+  public AccessTokenProvider getAccessTokenProvider() {
+    return accessTokenProvider;
+  }
+
+  /**
+   * Perform the unbonded deployment operations.
+   * Create the GCP credential provider chain to use
+   * when talking to GCP when there is no delegation token to work with.
+   * authenticating this client with GCP services, and saves it
+   * to {@link #accessTokenProvider}
+   *
+   * @throws IOException any failure.
+   */
+  public AccessTokenProvider deployUnbonded()
+      throws IOException {
+    Preconditions.checkState(!isBoundToDT(),
+        "Already Bound to a delegation token");
+    logger.atInfo().log("No delegation tokens present: using direct authentication");
+    accessTokenProvider = tokenBinding.deployUnbonded();
+    return accessTokenProvider;
+  }
+
+  /**
+   * Attempt to bind to any existing DT, including unmarshalling its contents
+   * and creating the GCP credential provider used to authenticate
+   * the client.
+   *
+   * If successful:
+   * <ol>
+   *   <li>{@link #boundDT} is set to the retrieved token.</li>
+   *   <li>{@link #decodedIdentifier} is set to the extracted identifier.</li>
+   *   <li>{@link #accessTokenProvider} is set to the credential
+   *   provider(s) returned by the token binding.</li>
+   * </ol>
+   * If unsuccessful, {@link #deployUnbonded()} is called for the
+   * unbonded codepath instead, which will set
+   * {@link #accessTokenProvider} to its value.
+   *
+   * This means after this call (and only after) the token operations
+   * can be invoked.
+   *
+   * @throws IOException selection/extraction/validation failure.
+   */
+  public void bindToAnyDelegationToken() throws IOException {
+    Preconditions.checkState(accessTokenProvider == null, E_ALREADY_DEPLOYED);
+    Token<AbstractGCPTokenIdentifier> token = selectTokenFromFSOwner();
+    if (token != null) {
+      bindToDelegationToken(token);
+    } else {
+      deployUnbonded();
+    }
+    if (accessTokenProvider == null) {
+      throw new DelegationTokenIOException("No AccessTokenProvider"
+          + " created by Delegation Token Binding "
+          + tokenBinding.getKind());
+    }
+  }
+
+  /**
+   * This is a test-only back door which resets the state and binds to
+   * a token again.
+   * This allows an instance of this class to be bonded to a DT after being
+   * started, so avoids the need to have the token in the current user
+   * credentials. It is package scoped so as to only be usable in tests
+   * in the same package.
+   *
+   * Yes, this is ugly, but there is no obvious/easy way to test token
+   * binding without Kerberos getting involved.
+   * @param token token to decode and bind to.
+   * @throws IOException selection/extraction/validation failure.
+   */
+  void resetTokenBindingToDT(final Token<AbstractGCPTokenIdentifier> token)
```

Comment:
Could you clarify if this is really needed, and if it is, how to use it?

Reply:
This is solely for testing support. If it's deemed unnecessary, then it can be removed without affecting actual functionality.